tests/periph_dac_play/main.c

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (C) 2020 Beuth Hochschule für Technik Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     tests
+ * @{
+ *
+ * @file
+ * @brief       DAC (audio) test application
+ *
+ * @author      Benjamin Valentin <benpicco@beuth-hochschule.de>
+ *
+ * @}
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "board.h"
+#include "thread.h"
+
+#include "msg.h"
+#include "mutex.h"
+#include "periph/dac.h"
+#include "periph/gpio.h"
+#include "periph/timer.h"
+#include "xtimer.h"
+
+#include "blob/hello.raw.h"
+
+#ifndef BTN0_INT_FLANK
+#define BTN0_INT_FLANK  GPIO_RISING
+#endif
+
+#ifndef DAC_CHAN
+#define DAC_CHAN DAC_LINE(0)
+#endif
+
+#ifndef ENABLE_GREETING
+#define ENABLE_GREETING (1)
+#endif
+
+#define DAC_BUF_SIZE (2048)
+
+static uint8_t buf[2][DAC_BUF_SIZE];
+
+static int32_t isin(int32_t x)
+{
+    // S(x) = x * ( (3<<p) - (x*x>>r) ) >> s
+    // n : Q-pos for quarter circle             13
+    // A : Q-pos for output                     12
+    // p : Q-pos for parentheses intermediate   15
+    // r = 2n-p                                 11
+    // s = A-1-p-n                              17
+    static const int qN = 13,
+                     qA = 12,
+                     qP = 15,
+                     qR = 2 * qN - qP,
+                     qS = qN + qP + 1 - qA;
+
+    x <<= (30 - qN);            // shift to full s32 range (Q13->Q30)
+
+    if ((x ^ (x << 1)) < 0) {   // test for quadrant 1 or 2
+        x = (1U << 31) - x;
+    }
+
+    x >>= (30 - qN);
+
+    return x * ((3 << qP) - (x * x >> qR)) >> qS;
+}
+
+static void _unlock(void *arg)
+{
+    mutex_unlock(arg);
+}
+
+static void _play_sin_sample(const dac_cfg_t *cfg, uint8_t b, uint8_t pitch)
+{
+    for (uint16_t i = 0; i < DAC_BUF_SIZE; ++i) {
+        buf[b][i] = (isin(i << pitch) + 4096) >> 5;
+    }
+
+    dac_play(buf[b], DAC_BUF_SIZE, cfg);
+    mutex_lock(cfg->cb_arg);
+}
```

Comment:
Wouldn't a constant buffer in ROM containing one period of the sine be much easier to use? By specifying the sample rate you should be able to generate different notes.

Reply:
Yea I also should clean this up.
My intention was to have an example for the callback function, for generating sound on the fly.
(Although usually the callback would rather be used to read from a file)

Reply:
Only using a single buffer now, not racing the playback anymore.